File: sample-app/src/main/scala/pl/newicom/jobman/test/ThreadFactoryBuilder.scala
package pl.newicom.jobman.test

import java.util.Locale
import java.util.concurrent.{Executors, ThreadFactory}
import java.util.concurrent.atomic.AtomicLong

object ThreadFactoryBuilder {

  def build(builder: ThreadFactoryBuilder): ThreadFactory = {
    val nameFormat                                                        = builder.nameFormat
    val daemon: Option[Boolean]                                           = builder.daemon
    val priority: Option[Int]                                             = builder.priority
    val uncaughtExceptionHandler: Option[Thread.UncaughtExceptionHandler] = builder.uncaughtExceptionHandler
    val backingThreadFactory =
      if (builder.backingThreadFactory != null) builder.backingThreadFactory else Executors.defaultThreadFactory

    val count = if (nameFormat != null) new AtomicLong(0) else null

    runnable: Runnable =>
      {
        val thread = backingThreadFactory.newThread(runnable)
        if (nameFormat != null) thread.setName(format(nameFormat, count.getAndIncrement))
        daemon.foreach(thread.setDaemon)
        priority.foreach(thread.setPriority)
        uncaughtExceptionHandler.foreach(thread.setUncaughtExceptionHandler)
        thread
      }
  }
  private def format(format: String, args: Any*) = String.format(Locale.ROOT, format, args)
}

final class ThreadFactoryBuilder() {
  private var nameFormat: String                                                = _
  private var daemon: Option[Boolean]                                           = None
  private var priority: Option[Int]                                             = None
  private var uncaughtExceptionHandler: Option[Thread.UncaughtExceptionHandler] = None
  private var backingThreadFactory: ThreadFactory                               = _

  def setNameFormat(nameFormat: String): ThreadFactoryBuilder = {
    val unused = ThreadFactoryBuilder.format(nameFormat, 0)
    this.nameFormat = nameFormat
    this
  }

  def setDaemon(daemon: Boolean): ThreadFactoryBuilder = {
    this.daemon = Option(daemon)
    this
  }

  def setPriority(priority: Int): ThreadFactoryBuilder = {
    this.priority = Option(priority)
    this
  }

  def setUncaughtExceptionHandler(uncaughtExceptionHandler: Thread.UncaughtExceptionHandler): ThreadFactoryBuilder = {
    this.uncaughtExceptionHandler = Option(uncaughtExceptionHandler)
    this
  }

  def setThreadFactory(backingThreadFactory: ThreadFactory): ThreadFactoryBuilder = {
    this.backingThreadFactory = backingThreadFactory
    this
  }

  def build: ThreadFactory = ThreadFactoryBuilder.build(this)
}

File: job-man-view-sql/src/main/java/pl/newicom/jobman/view/sql/ThreadFactoryBuilder.java
package pl.newicom.jobman.view.sql;

import java.lang.Thread.UncaughtExceptionHandler;
import java.util.Locale;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicLong;

public final class ThreadFactoryBuilder {
	private String nameFormat = null;
	private Boolean daemon = null;
	private Integer priority = null;
	private UncaughtExceptionHandler uncaughtExceptionHandler = null;
	private ThreadFactory backingThreadFactory = null;

	public ThreadFactoryBuilder() {}

	public ThreadFactoryBuilder setNameFormat(String nameFormat) {
		String unused = format(nameFormat, 0); // fail fast if the format is bad or null
		this.nameFormat = nameFormat;
		return this;
	}

	public ThreadFactoryBuilder setDaemon(boolean daemon) {
		this.daemon = daemon;
		return this;
	}

	public ThreadFactoryBuilder setPriority(int priority) {
		this.priority = priority;
		return this;
	}

	public ThreadFactoryBuilder setUncaughtExceptionHandler(
			UncaughtExceptionHandler uncaughtExceptionHandler) {
		this.uncaughtExceptionHandler = uncaughtExceptionHandler;
		return this;
	}

	public ThreadFactoryBuilder setThreadFactory(ThreadFactory backingThreadFactory) {
		this.backingThreadFactory = backingThreadFactory;
		return this;
	}

	public ThreadFactory build() {
		return doBuild(this);
	}

	private static ThreadFactory doBuild(ThreadFactoryBuilder builder) {
		final String nameFormat = builder.nameFormat;
		final Boolean daemon = builder.daemon;
		final Integer priority = builder.priority;
		final UncaughtExceptionHandler uncaughtExceptionHandler = builder.uncaughtExceptionHandler;
		final ThreadFactory backingThreadFactory =
				(builder.backingThreadFactory != null)
						? builder.backingThreadFactory
						: Executors.defaultThreadFactory();
		final AtomicLong count = (nameFormat != null) ? new AtomicLong(0) : null;
		return new ThreadFactory() {
			@Override
			public Thread newThread(Runnable runnable) {
				Thread thread = backingThreadFactory.newThread(runnable);
				if (nameFormat != null) {
					thread.setName(format(nameFormat, count.getAndIncrement()));
				}
				if (daemon != null) {
					thread.setDaemon(daemon);
				}
				if (priority != null) {
					thread.setPriority(priority);
				}
				if (uncaughtExceptionHandler != null) {
					thread.setUncaughtExceptionHandler(uncaughtExceptionHandler);
				}
				return thread;
			}
		};
	}

	private static String format(String format, Object... args) {
		return String.format(Locale.ROOT, format, args);
	}
}